Clean up lots of win32 crustiness

tari opened the pull request:
This builds on top of #74 to remove a lot of old crusty patterns that mostly relate to targeting Windows as a primary platform. Using CMake, vcpkg and C++17 features:

 * all uses of raw win32 APIs are removed in favor of standard C++
 * the COM interface is removed (nobody seems to care about it and making it work with cmake seems hard)
 * app signing on Windows doesn't depend on random prebuilt .lib files anymore (GMP will be automatically built from source if needed)
 * dependency on OpenSSL is removed in favor of shipping a public-domain implementation of MD5

tari committed:
Pass required string for ERR_INVALID_OPERANDS

When unable to parse an operand to something that appears
to be a macro, ensure the required string for the error message is
passed. Failing to do so causes undefined behavior when the error
message is formatted, and reliably crashes.

Fixes #71.

tari committed:
Support building with cmake

This should permit unification of the makefile and msbuild projects.

tari committed:
Update github CI to use cmake

tari committed:
Rip out a lot of windows idioms

This replaces TCHAR with explicit unicode and swaps windows-specific
types like DWORD with standard equivalents, as well as removing the COM
interface.

tari committed:
Rip out stdafx.h

This is a peculiar windows idiom for precompiled headers
that isn't important to us.

tari committed:
Use vcpkg to get libraries on Windows

tari committed:
Delete obsolete .travis.yml

Travis hasn't been truly free for open-source projects since 2021, and
is no longer used by this project.

tari committed:
Remove support for building with Make

CMake should be used going forward.

tari committed:
Merge pull request #1 from tari/cmake

Support building with cmake

tari committed:
Replace bare windows/POSIX APIs with c++17

c++17's std::filesystem allows us to get rid of some mess using either
bare win32 or POSIX APIs depending on the target platform. The code is
easier to understand and shorter.

tari committed:
Remove use of std::binary_function

This is unneeded after c++11, and broken from c++17.

Fixes https://github.com/alberthdev/spasm-ng/issues/73

tari committed:
Silence MSVC unsafe function warnings

"consider using sprintf_s instead" and the like is not helpful because
it's not portable.

tari committed:
Include a public-domain MD5 implementation

This removes the dependency on the Windows cryptography libraries and
OpenSSL, at the cost of including a little bit more code.

tari committed:
tests: use source-relative paths for names

Take the relative path of each automatically-discovered test case,
rather than the absolute path returned by file(GLOB). This makes the
test names much simpler.

tari committed:
Add pkgconf dependency for Windows

This allows cmake to use PKG_CONFIG without any other installs.
https://github.com/microsoft/vcpkg/issues/17481

calc84maniac commented on preop.cpp:
```diff
@@ -405,43 +406,21 @@ char *handle_preop_define (const char *ptr) {
-#endif
+char *full_path (const char *filename_bytes) {
+	std::error_code fs_err;
+	const std::filesystem::path filename(filename_bytes);
```
This doesn't do UTF-8 conversion, unlike the code you're replacing. Since you're building C++17, you should use std::filesystem::u8path() to generate the path object.

tari replied:
The diff on this commit is a little misleading, since I added the UTF-8 handling in d56556d3e1d11c283592518b6a6a52799eab004f but don't think that was actually correct. The `basic_string` path constructor seems correct because we haven't actually checked what the encoding of our bytestrings is.

I think this means it'll end up using the ANSI encoding, but I'm not sure there's a better solution and if the process is running with the UTF-8 codepage then we get exactly what we want.

calc84maniac replied:
Ah, that makes sense. I think the ideal "correct" solution is accepting wide string CLI arguments and converting them to UTF-8, but that brings back some crustiness. Using UTF-8 for filenames specified in source files does seem like the most consistent thing to do, however...

tari replied:
Maybe the best thing to do is embed an application manifest in Windows builds that sets UTF-8 mode (`activeCodePage` = `UTF-8`), then we can ignore the rest of the Window oddities.

tari replied:
Added an application manifest in 0e961dac927077d45ec67c4b00a57414785d33c2 that requests UTF-8 mode that I think solves this, at least for sufficiently recent versions of Windows.

calc84maniac commented on bitmap.cpp:
```diff
@@ -14,7 +14,7 @@ typedef uint16_t UINT;
 typedef uint16_t WORD;
 // Defined in stdafx.h
-//typedef uint32_t DWORD;
+//typedef uint32_t int32_t;
```
Why change all these unsigned values to signed? It especially doesn't make sense when dealing with RGB color values and sizes

tari replied:
That was a mistake. Changing some instances in bitmap.cpp exposed that it uses windows type names extensively and cleaning that up led to a number of other uses of windows type names that I've changed in dadc04c8b99b519355997eb7e26b4c20244edf33.

tari committed:
Remove more windows type names

This avoids including Windows.h except where actually required, and deletes
more uses of windows type names like WORD. bitmap.cpp is normalized to use
the windows names because it uses them extensively, but it no longer includes
Windows.h. Some former DWORDs that were incorrectly converted to int32_t are
changed to be unsigned again.

tari committed:
Add a Windows application manifest

In particular this requests UTF-8 mode for versions of Windows that support
using the UTF-8 codepage. This is useful because it allows us to use bytestrings
and assume they're always UTF-8, without needing to deal with the mess that is
Windows wide characters and UCS-2 being completely unusable for non-BMP characters.

tari committed:
Silence MSVC warnings about spectre mitigations

We don't actually care, but it's still probably useful to enable -Wall.

tari committed:
tests: expect UTF-8 output

If textual output ends up containing garbage, we'd like to know.
This will display badly-encoded data if we find it, making
that somewhat easier to debug.

tari committed:
Add a macos target to CI builds
